csharp/src/Apache.Arrow/Arrays/DecimalArray.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Apache.Arrow.Types;
+
+namespace Apache.Arrow
+{
+    public class DecimalArray : PrimitiveArray<decimal>
```

Comment:
I don't think the C# in-memory format of decimal is the same layout as the Arrow decimal format, is it? I'm trying to find the documentation on the memory layout, the only thing I can find is in the Schema.fbs:

https://github.com/apache/arrow/blob/e676aeacb2c6597c063faeed20d6142b0df938d8/format/Schema.fbs#L176-L185

The C# decimal layout is different:

https://docs.microsoft.com/en-us/dotnet/api/system.decimal?view=net-5.0#remarks

> The binary representation of a Decimal value consists of a 1-bit sign, a 96-bit integer number, and a scaling factor used to divide the 96-bit integer and specify what portion of it is a decimal fraction. The scaling factor is implicitly the number 10, raised to an exponent ranging from 0 to 28.

Reply:
The documentation is trying to say that the Decimal value is either a 128-bit or a 256-bit twos complement integer.  The metadata of the type determines the scaling factor of the integer.  So yes, it sounds like they are quite different.

Reply:
Thanks @emkornfield. One thing that I don't quite understand is the `precision` field in metadata. The data is fixed sized right? So each value will always take up 128 or 256 bits. But if I set the precision lower than `38` for a Decimal128, what exactly does that mean? Does it mean to ignore digits at the beginning of the integer? For example:

```
Precision = 3
Scale = 0
Data Value = 0b010111011100   (1,500 in binary)
```

I assume this number means `500`, and the leading `1,000` gets truncated.

Reply:
Yes, we ignore most significant bits of the integer (that would correspond to higher necessary precision).  I'm sure some code might break if those bits are populated.  But for instance we use this fact when writing to parquet: https://github.com/apache/arrow/blob/0e8de085f4afd3c9aa9c9ed05405a6c6ef8c360a/cpp/src/parquet/column_writer.cc